### PR TITLE
ci: add CircleCI integration and ci-check script.

### LINKED ISSUE #10: add CI integration and run format checker, linter, etc

### PR DESCRIPTION
**Pull request by fabiosantoscode**
I haven't fixed the linting or typing errors so I've commented out that part.

Also added a `yarn check-all` so we can run `prettier --check` instead of `prettier --write` to ensure all our files are merged to master properly linted.

Fixes #10

### CODE REVIEW

**Comment by fabiosantoscode** [comment]
@shcheklein I don't have permissions but we probably need to enable CircleCI for this repository.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      - install
+      - run: yarn ci-check
+      - cache
+
+workflows:
+  version: 2
+
+  commit:
+    jobs:
+      - test
```
Comment (fabiosantoscode) [inline .circleci/config.yml]: This is largely a copy of what we have in the dvc.org repo, trimmed down to what we need.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Javascript Node CircleCI 2.0 configuration file
```
Comment (shcheklein) [inline .circleci/config.yml]: I think you need more stuff from the dvc.org's file - restoring the branch for example.

Comment (fabiosantoscode) [reply .circleci/config.yml]: I saw that it was a workaround for tools which compare HEAD to master. None of the tools we're using do that.

File: package.json [hunk 3]
```diff
@@ -12,7 +12,9 @@
     "typecheck": "tsc --noEmit",
     "lint-ts": "tslint --project .",
     "lint-css": "stylelint \"src/**/*.css\"",
-    "format-all": "prettier --write './**/*.{js,jsx,md,tsx,ts,json}'"
+    "format-all": "prettier --write './**/*.{js,jsx,md,tsx,ts,json}'",
+    "check-all": "prettier --check './**/*.{js,jsx,md,tsx,ts,json}'",
+    "ci-check": "yarn check-all && yarn lint-css # TODO lint and type files && yarn lint-css && yarn typecheck"
```
Comment (shcheklein) [inline package.json]: please, don't include TODO - create a ticket if there is something to be fixed

Comment (shcheklein) [reply package.json]: we should not run the full check - it's slower and is not related to PR,  I think we always run only files from the PR

Comment (shcheklein) [reply package.json]: and yep, we need all other checks we have

Comment (fabiosantoscode) [reply package.json]: Got it

Comment (fabiosantoscode) [reply package.json]: The checkers we are using are pretty fast. But I'll try to check only modified files.

**Comment by fabiosantoscode** [comment]
Ok, now only files which have changed between master and the PR will be linted. I didn't find any package which did this out of the box so I made a script for it. And because we're already configuring `lint-staged` I thought I might as well read its configuration directly.

The script is essentially `git diff master --name-only | xargs <linter>`, and selects which linter to use according to file globs.

The output looks like this:

```
LINT-CHANGED> yarn prettier --write scripts/lint-changed.js
yarn run v1.17.3
$ /home/fabio/devel/iterative/blog/node_modules/.bin/prettier --write scripts/lint-changed.js
scripts/lint-changed.js 59ms
Done in 0.26s.

LINT-CHANGED> yarn lint-ts scripts/lint-changed.js
yarn run v1.17.3
$ tslint --project . scripts/lint-changed.js
Done in 1.89s.
```

**Comment by shcheklein** [comment]
@fabiosantoscode what are the checks on the dvc.org that required clearing the master? Trying to understand the logic behind stuff. Current solution looks too complicated - and there are a few ways to make it simpler, but first I'd like to restore the full picture of what is going on on dvc.org and why.

**Comment by shcheklein** [comment]
so, looking closely into this I think we don't need this restore master procedure anymore (after @jorgeorpinel changed it from pretty-quick to the direct prettier call.

let's confirm that dvc.org actually runs all the checks (except links) for the whole project and then cleanup that master branch restoration on both projects and get back to your previous simple logic here. Except that we need to run all checks - the same way we do for dvc.org. Sorry for the confusion around checks on the dvc.org end.

**Comment by fabiosantoscode** [comment]
@shcheklein ok, I'll go back to that.

No worries, it should be really fast - many times faster than the netlify deploy - even if we check all files.

I'll revert the changes.

Do you think you need to enable circleCI in this repository? I'd like to see if this actually runs there.

**Comment by fabiosantoscode** [comment]
What's happening on dvc.org is a full repo check every time.

**Comment by fabiosantoscode** [comment]
@shcheklein I suggest we do the full-project linting as part of https://github.com/iterative/blog/issues/101. It also includes adding the missing checks where I had that TODO comment before. But I can do it in this PR if you prefer.

File: package.json [hunk 4]
```diff
@@ -12,7 +12,9 @@
     "typecheck": "tsc --noEmit",
     "lint-ts": "tslint --project .",
     "lint-css": "stylelint \"src/**/*.css\"",
-    "format-all": "prettier --write './**/*.{js,jsx,md,tsx,ts,json}'"
+    "format-all": "prettier --write './**/*.{js,jsx,md,tsx,ts,json}'",
+    "check-all": "prettier --check './**/*.{js,jsx,md,tsx,ts,json}'",
+    "ci-check": "yarn check-all && yarn lint-css"
```
Comment (shcheklein) [inline package.json]: it does not look like _all_?  we have lint-ts ... (and yes, let's fix the complaints)

also, it's better to keep steps in the CircleCi separate - easier to navigate and see what specific check fails and you can avoid introducing this "ci-check" here - it's not needed

**Comment by shcheklein** [comment]
please, rebase - there is a conflict

**Comment by shcheklein** [comment]
I've enabled CircleCI, but I'm not sure I can it for the fork

**Comment by shcheklein** [comment]
I suppose we'll have to merge it and see how it goes

**Comment by fabiosantoscode** [comment]
@shcheklein squashed it up and added the typecheck before each commit. Typechecking individual files didn't work but the full project check is less than 3 seconds.

**Comment by shcheklein** [comment]
@fabiosantoscode what is needed for? I mean typecheck. I'm not familiar but don't we run `tsc` to compile code anyway?

**Comment by fabiosantoscode** [comment]
@shcheklein the typecheck is for verifying that the types check out. For example, it checks that you're not passing numbers to a function that takes strings.

**Comment by shcheklein** [comment]
@fabiosantoscode merged! check the results here - https://circleci.com/gh/iterative/blog/2

**Pull request merged**